drivers: uart: save individual uart instances interrupts in PM Action

davidmissael opened the pull request:
The PM action saves the interrupts enabled using a global variable called usart_intenset. The problem is when you have multiple uart instances that have different configurations, when entering/exiting from a power level, only one configuration is saved and will be applied to all uart instances when exiting from a power level. We need to keep this setting individual for each instance. To do this, usart_intenset was added as a new element to mcux_flexcomm_data structure. In this way, each uart instance will keep/restore its own setting.

davidmissael committed:
drivers: uart: save interrupts in PM Action

The PM action saves the interrupts enabled using a
global variable called usart_intenset. The problem is
when you have multiple uart instances that have different
configurations, when entering/exiting from a power level,
only one configuration is saved and will be applied to all
uart instances when exiting from a power level. We need to
keep this setting individual for each instance. To do this,
usart_intenset was added as a new element to mcux_flexcomm_data
structure. In this way, each uart instance will keep/restore
each own setting.

Signed-off-by: Missael Maciel <davidmissael.maciel@nxp.com>

Albort12138 commented on drivers/serial/uart_mcux_flexcomm.c:
```diff
@@ -1217,14 +1220,18 @@ static int mcux_flexcomm_pm_action(const struct device *dev, enum pm_device_acti
 	case PM_DEVICE_ACTION_TURN_OFF:
-		usart_intenset = USART_GetEnabledInterrupts(config->base);
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+		data->usart_intenset = USART_GetEnabledInterrupts(config->base);
```
Could you share your use case? In what scenario would PM_DEVICE_ACTION_TURN_OFF be executed? Why, during TURN_OFF, do you only save the enabled interrupts instead of performing any operations related to turning off?

davidmissael replied:
Hi @Albort12138 PM_DEVICE_ACTION_TURN_OFF is being executed when the device enters in a lowpower mode. The specific case for the RW612, when entering in PM3, the microcontoller does not have registers retention (only SRAM), so any configuration at any peripheral will be lost, that's why this configuration has to be saved during PM_DEVICE_ACTION_TURN_OFF and restored at PM_DEVICE_ACTION_TURN_ON. This is the only operation we need to save in this action to recover the uart while exiting from low power mode. **If you take a look on the driver that is currently  implemented** this is the only operation performed, actually I'm not modifying that logic, this PR just tries to save this context individually per each uart instance defined in the application instead of having one single global variable that retains that configuration and restores all the other ones. We need one context per instance.
@dleach02 @decsny @mmahadevan108 since you are the driver's maintainers, is there anything else we can add to the @Albort12138's question?

davidmissael committed:
drivers: uart: save interrupts in PM Action

Removed usart_intenset attribute from conditional compilation
in mcux_flexcomm_data structure since this parameter needs to
be saved/restored independently if the interrupts are enabled
or not based on the PR feedback

Signed-off-by: Missael Maciel <davidmissael.maciel@nxp.com>
